File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:


        def gcd(a, b):
            while b:
                a, b = b, a % b
            return a

        
        if not head or not head.next:
            return head
        p = head  

        while p and p.next:
            q = p.next  
            val = gcd(p.val, q.val)
            new_node = ListNode(val)
            new_node.next = q
            p.next = new_node
            p = q
        
        return head

File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int gcd(int a, int b){
        while (b!=0){
            int temp = a % b;
            a = b;
            b = temp;
        }

        return a;
    }

    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }

        ListNode p = head;
        ListNode q = head.next;
        while(p!=null && q!=null){
            int val = gcd(p.val, q.val);
            ListNode n = new ListNode(val);
            n.next = q;
            p.next = n;
            p = q;
            q = q.next;
        }

        return head;
        
    }
}